File: progra3-0583-equipo5-frontend/Frontend_C#/SoftPac/SoftPacBusiness/FacturasBO.cs
using SoftPac.Model;
using SoftPac.Persistance.DAO;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SoftPac.Business
{
    public class FacturasBO
    {
        FacturasDAO facturaDAO;

        public FacturasBO()
        {
            this.facturaDAO = new FacturasDAOImpl();
        }
        public BindingList<FacturasDTO> ListarTodos()
        {
            BindingList<FacturasDTO> facturas = (BindingList<FacturasDTO>)this.facturaDAO.ListarTodos();

            foreach (var factura in facturas)
            {
                System.Diagnostics.Debug.WriteLine("Factura encontrada: " + factura.NumeroFactura);
            }

            return facturas;
        }

        public int Eliminar(FacturasDTO factura)
        {
            return this.facturaDAO.eliminarLogico(factura);
        }

        public int Eliminar(int facturaId, UsuariosDTO usuario)
        {
            FacturasDTO factura = new FacturasDTO();
            factura.FacturaId = facturaId;
            factura.UsuarioEliminacion= usuario;
            factura.FechaEliminacion = DateTime.Now;
            return Eliminar(factura);
        }

        public int InsertarDetalle(DetallesFacturaDTO detalle)
        {
            return this.facturaDAO.Modificar(detalle.Factura);
        }

        public int ModificarDetalle(DetallesFacturaDTO detalle)
        {
            return this.facturaDAO.Modificar(detalle.Factura);
        }

        public int EliminarDetalle(DetallesFacturaDTO detalle,UsuariosDTO usuario)
        {
            detalle.UsuarioEliminacion = usuario;
            detalle.FechaEliminacion = DateTime.Now;
            return this.facturaDAO.Modificar(detalle.Factura);
        }

        public FacturasDTO ObtenerPorId(int facturaId)
        {
            return facturaDAO.ObtenerPorId(facturaId);
        }

        public int Modificar(FacturasDTO factura)
        {
            return facturaDAO.Modificar(factura);
        }

        public int Insertar(FacturasDTO factura)
        {
            return facturaDAO.Insertar(factura);
        }

    }
}

File: Backend_Java/SoftPacBusiness/src/main/java/pe/edu/pucp/softpac/bo/FacturasBO.java
package pe.edu.pucp.softpac.bo;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Date;
import pe.edu.pucp.softpac.dao.FacturasDAO;
import pe.edu.pucp.softpac.daoImpl.FacturasDAOImpl;
import pe.edu.pucp.softpac.model.AcreedoresDTO;
import pe.edu.pucp.softpac.model.DetallesFacturaDTO;
import pe.edu.pucp.softpac.model.FacturasDTO;
import pe.edu.pucp.softpac.model.MonedasDTO;
import pe.edu.pucp.softpac.model.UsuariosDTO;

public class FacturasBO {
    
    private FacturasDAO facturasDAO;
    
    
    public FacturasBO(){
        facturasDAO = new FacturasDAOImpl();
   
    }
    
    public ArrayList<FacturasDTO> listarTodos(){
        return (ArrayList<FacturasDTO>) facturasDAO.listarTodos();
    }
    
    public Integer eliminarLogico(FacturasDTO factura, UsuariosDTO usuarioActual){
        factura.setFecha_eliminacion(new Date());
        factura.setUsuario_eliminacion(usuarioActual);
        
        return facturasDAO.eliminarLogico(factura);
    }
    
    public Integer insertarDetalle(DetallesFacturaDTO detalle){
        detalle.getFactura().addDetalle_Factura(detalle);
        return this.facturasDAO.modificar(detalle.getFactura());
    }
    
    public Integer modificarDetalle(DetallesFacturaDTO detalle){
        detalle.getFactura().addDetalle_Factura(detalle);
        return this.facturasDAO.modificar(detalle.getFactura());
    }
    
    public Integer eliminarDetalle(DetallesFacturaDTO detalle, UsuariosDTO usuarioActual){
        detalle.setFecha_eliminacion(new Date());
        detalle.setUsuario_eliminacion(usuarioActual);
        detalle.getFactura().addDetalle_Factura(detalle);
        return this.facturasDAO.modificar(detalle.getFactura());
    }   
    
    public Integer insertar(FacturasDTO factura){
        
        return this.facturasDAO.insertar(factura);
        
    }
    
    public Integer modificar(FacturasDTO factura){
        if(factura.getDetalles_Factura()!=null){
            for(DetallesFacturaDTO detalle : factura.getDetalles_Factura()){
                detalle.setFactura(factura);
            }
        }
        return this.facturasDAO.modificar(factura);
                
    }
    
    public ArrayList<FacturasDTO> listarPendientes(){
        ArrayList<FacturasDTO> facturasPendientes = new ArrayList<>();
        ArrayList<FacturasDTO> facturasGenerales = (ArrayList<FacturasDTO>) facturasDAO.listarTodos();
        for(FacturasDTO factura : facturasGenerales){
            if(factura.getEstado().equals("Pendiente")){
                facturasPendientes.add(factura);
            }
        }
        
        return facturasPendientes;
        
    }
 
    public FacturasDTO obtenerPorId(Integer factura_id){
        if (factura_id == null || factura_id <= 0) {
            throw new IllegalArgumentException("El ID de la factura es obligatorio");
        }
        
        FacturasDTO factura = new FacturasDTO();
        factura.setFactura_id(factura_id);
        return this.facturasDAO.obtenerPorId(factura_id);
    }
    
    public ArrayList<FacturasDTO> listarPendientesPorCriterios(Integer paisId, Date fechaLimite) {
    ArrayList<FacturasDTO> resultado = new ArrayList<>();

        ArrayList<FacturasDTO> facturasGenerales = (ArrayList<FacturasDTO>) facturasDAO.listarFiltros();

        for (FacturasDTO factura : facturasGenerales) {
            if (
                factura != null
                && "Pendiente".equals(factura.getEstado())
                && factura.getAcreedor() != null
                && factura.getAcreedor().getPais() != null
                && factura.getAcreedor().getPais().getPais_id().equals(paisId)
                && factura.getFecha_limite_pago() != null
                && ( !factura.getFecha_limite_pago().after(fechaLimite) )
            ) {
                resultado.add(factura);
            }
        }
        return resultado;
    }
    
    public void vencerFacturas() {
        
        try {
            ArrayList<FacturasDTO> facturas = listarTodos();
        
            for (FacturasDTO factura : facturas)
            {
                if (factura.getFecha_limite_pago().before(new Date()) && "Pendiente".equals(factura.getEstado()))
                {
                    factura.setEstado("Vencida");
                    modificar(factura);
                }
            }
            
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
        }
    }
    
    
}